parlai/agents/transformer/modules/encoder.py

```diff
@@ -20,12 +22,75 @@
     MultiHeadAttention,
     TransformerFFN,
 )
+from parlai.agents.transformer.modules.interfaces import ComponentSpec, TComponent
 from parlai.core.opt import Opt
 from parlai.utils.misc import warn_once
 from parlai.utils.torch import PipelineHelper
 
 
-class TransformerEncoder(nn.Module):
+class TransformerEncoderLayer(nn.Module, TComponent):
+    """
+    Implements a single Transformer encoder layer.
+    """
+
+    @dataclass
+    class Manifest(TComponent.Manifest):
+        self_attention: Type[MultiHeadAttention] = MultiHeadAttention
```

Comment:
why are these not component specs?

Reply:
my guess is MHA is considered bottom-level module, with manifest  = None (no swappable components)

Reply:
Yep, what Jing said. `MultiHeadAttention` isn't a `TComponent` at this point because there aren't any swappable components within it.

Maybe there's a better name for `TComponent` and `ComponentSpec` that makes this clearer. Like... `TComponent` -> `Composable`, `ComponentSpec` -> `ComposeKit`?

Idk I'm very open to suggestions. 

The idea I'm trying to convey is:
- A "TComponent"/"Composable"/"Modular" component is a piece of the model with modular pieces that can be substituted with anything that conforms to the same `__init__` and `forward` as the original piece.
- To fully specify the "TComponent"/"Composable" component, then, you need to specify both the base class and the class of the pieces that will be plopped into the various slots within it.
- The "TComponent"/"Composable" components come with a default set of pieces so you don't have to think about this until you want to change something.
- A "ComponentSpec"/"ComposeKit" is a fully specified "TComponent"/"Composable", in that it describes both the base class and all the pieces that fit into its slots.

Reply:
"Mad Libs" are a fantastic analogy and it's kind of cool that it hints at the idea that code and natural language are interchangeable (a sequence is a sequence). But idk if `PhrasalTemplate`, `Word`, and `Phrase` are any more intuitive than `TComponent`, `Module`, and `ComponentSpec` lol

https://en.wikipedia.org/wiki/Phrasal_template

Reply:
I think i prefer "Composable"/"Modular" to component, as it's not immediately clear why a component of a transformer may not be composable (e.g. in the MHA case). Perhaps we can specify even further to differentiate those components. In that regard, you might have something like:

"ModularComponent": A piece of the model with modular pieces etc. etc.
"StaticComponent": A piece of the model that does not have modular pieces etc. etc.
"ModularComponentSpec": The specification of the modular component

Does that make sense in this setup?

Reply:
I like that naming better! Changing it now.

"StaticComponent" might be implied and not literal, though. Unless we can figure out a way to use a "StaticComponent" type to _enforce_ that an nn.Module not be composed.

Reply:
I'm ok with it just being semantic in meaning (as Python makes it hard to really enforce anything lol)

Reply:
Okay! I added a `StaticComponent` class that basically just acts as a comment.